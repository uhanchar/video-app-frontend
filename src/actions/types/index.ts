export * from 'actions/types/app-types';
export * from 'actions/types/video-types';
