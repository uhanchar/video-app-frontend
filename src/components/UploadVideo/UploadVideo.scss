@import "src/constants";

.upload-video {
  display: flex;
  flex-direction: column;
  align-items: center;
  flex-grow: 1;
  height: calc(100vh - #{ $header-height });
  padding: 1.875rem 1.25rem;
  box-sizing: border-box;

  > .title {
    margin-bottom: 6rem;
    font-size: 1.25rem;
  }

  > .form {
    width: 100%;

    > .container {

      > .upload-controls {
        display: flex;
        justify-content: center;

        > .filename {
          display: flex;
          justify-content: center;
          align-items: center;
          height: 3rem;
          color: #90a0b7;
          font-size: 1rem;
        }

        > .error {
          color: #e32020;
          font-size: 1.125rem;
          font-weight: 500;
        }

        > .input {
          display: none;
        }

        .progress {
          width: 100%;
        }
      }

      .submit-button {
        width: 100%;
        background: $color-success;
        color: $color-blank;

        &:disabled {
          background: #e3dede;
          color: #a09c9c;
        }
      }
    }
  }
}
