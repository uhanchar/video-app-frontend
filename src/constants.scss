// colors
$color-blank: #ffffff;
$color-success: #2ed47a;
$primary-color-black: #020003;

// sizes
$header-height: 4rem;
