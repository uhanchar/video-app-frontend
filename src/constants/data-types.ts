export type NullableType<T> = T | null;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type CustomAny = any;
