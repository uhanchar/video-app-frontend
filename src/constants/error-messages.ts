export const WRONG_FILE_FORMAT_MESSAGE = 'Wrong file format. Available formats are *.mp4, *.mkv, *.avi, *.wmv, *.mpeg4';
