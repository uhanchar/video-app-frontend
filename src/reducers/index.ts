import { combineReducers } from 'redux';
import AppReducer from 'reducers/app.reducer';

const rootReducer = combineReducers({
  app: AppReducer,
});

export default rootReducer;
