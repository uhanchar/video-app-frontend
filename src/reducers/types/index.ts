import { NullableType } from 'constants/data-types';

export interface IVideoItem {
  id: string;
  name: string;
  description: string;
  link: string;
  thumbnailLink: string;
  createdAt: Date;
}

export interface IAppState {
  videos: IVideoItem[];
}

export interface IVideoState {
  video: NullableType<IVideoItem>;
  isLoading: boolean;
  error: string;
}

export type RootState = {
  app: IAppState;
  video: IVideoState;
};
