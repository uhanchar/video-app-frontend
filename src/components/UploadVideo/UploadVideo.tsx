import React, { ChangeEvent, FC, SyntheticEvent, useCallback, useRef, useState } from 'react';
import { Button, Grid, LinearProgress, TextField } from '@material-ui/core';

import 'components/UploadVideo/UploadVideo.scss';
import { isValidFileSize, isValidFormat } from 'helpers/file-format-validation.helper';
import { ErrorMessages } from 'constants/error-messages.enum';
import { uploadVideo } from 'services/upload-video.service';
import UploadConfirmation from 'components/UploadConfirmation/UploadConfirmation';
import Error from 'components/Error/Error';

const UploadVideo: FC = () => {
  const [ fileName, setFileName ] = useState<string>('');
  const [ fileTitle, setFileTitle ] = useState<string>('');
  const [ fileDescription, setFileDescription ] = useState<string>('');
  const [ file, setFile ] = useState<File | null>(null);
  const [ error, setError ] = useState<string>('');
  const [ serverError, setServerError ] = useState('');
  const [ isLoading, setIsLoading ] = useState<boolean>(false);
  const [ loadingProgress, setIsLoadingProgress ] = useState<number>(0);
  const [ isConfirmationOpen, setIsConfirmationOpen ] = useState<boolean>(false);
  const fileInput = useRef<HTMLInputElement | null>(null);

  const onBrowseClick = () => {
    if (fileInput.current) {
      (fileInput.current! as HTMLInputElement).value = '';
    }
  };

  const handleFileChange = (event: ChangeEvent<HTMLInputElement>) => {
    setError('');
    setServerError('');

    const fileToUpload = event.target.files?.item(0);

    if (!fileToUpload) {
      return;
    }

    if (!isValidFormat(fileToUpload.name)) {
      setError(ErrorMessages.WRONG_FILE_FORMAT_MESSAGE);

      return;
    }

    if (!isValidFileSize(fileToUpload.size)) {
      setError(ErrorMessages.INVALID_FILE_SIZE);

      return;
    }

    setFileName(fileToUpload.name);
    setFile(fileToUpload);
  };

  const onFileTitleChange = (event: ChangeEvent<HTMLInputElement>) => {
    setFileTitle(event.target.value);
  };

  const onFileDescriptionChange = (event: ChangeEvent<HTMLInputElement>) => {
    setFileDescription(event.target.value);
  };

  const clearState = () => {
    setFileName('');
    setFileTitle('');
    setFileDescription('');
    setFile(null);
    setIsLoading(false);
    setIsLoadingProgress(0);
  };

  const handleFileDataSubmit = async (event: SyntheticEvent) => {
    event.preventDefault();
    setServerError('');
    setIsLoading(true);

    try {
      await uploadVideo({ fileName: fileTitle, fileDescription, file: file! }, setIsLoadingProgress);

      clearState();
      setIsConfirmationOpen(true);
    } catch (err) {
      setIsLoading(false);
      setIsLoadingProgress(0);

      const errorMessage = err?.response?.data;
      setServerError(errorMessage || ErrorMessages.GENERIS_SERVER_ERROR_MESSAGE);
    }
  };

  const closeConfirmation = useCallback(() => {
    setIsConfirmationOpen(false);
  }, []);

  const sendButtonDisabled = !!error || isLoading || !fileTitle.trim().length || !file;

  const renderError = (content: string) => (
    content && (
      <Error content={content} />
    )
  );

  const renderFileLoadingProgress = () => (
    isLoading && (
      <Grid className="upload-controls" item xs={12} data-testid="linear-progress">
        <Grid item xs={12} md={6} lg={4}>
          <LinearProgress className="progress" variant="determinate" value={loadingProgress} />
        </Grid>
      </Grid>
    )
  );

  return (
    <div className="upload-video">
      <div className="title">Upload video</div>

      <form className="form">
        <Grid container className="container" spacing={3} justify="center" alignItems="center">
          <Grid className="upload-controls" item xs={12}>
            <div className="filename" data-testid="file-name">{ fileName }</div>
          </Grid>

          { renderError(error) }

          { renderError(serverError) }

          { renderFileLoadingProgress() }

          <Grid className="upload-controls" item xs={12} md={6} lg={4}>
            <input
              id="contained-button-file"
              className="input"
              accept="video/*"
              type="file"
              multiple={false}
              ref={fileInput}
              disabled={isLoading}
              data-testid="file-input"
              onChange={handleFileChange}
            />

            <label htmlFor="contained-button-file">
              <Button
                variant="contained"
                color="primary"
                component="span"
                disabled={isLoading}
                onClick={onBrowseClick}
              >
                Upload
              </Button>
            </label>
          </Grid>

          <Grid container item xs={12} justify="center" alignItems="center">
            <Grid item xs={12} md={6} lg={4}>
              <TextField
                id="video-title"
                label="Title"
                variant="outlined"
                className="text-input"
                fullWidth
                required
                data-testid="file-title"
                value={fileTitle}
                onChange={onFileTitleChange}
              />
            </Grid>
          </Grid>

          <Grid container item xs={12} justify="center" alignItems="center">
            <Grid item xs={12} md={6} lg={4}>
              <TextField
                id="description"
                label="Description"
                variant="outlined"
                multiline
                fullWidth
                data-testid="file-description"
                className="text-input"
                value={fileDescription}
                onChange={onFileDescriptionChange}
              />
            </Grid>
          </Grid>

          <Grid container item xs={12} justify="center" alignItems="center">
            <Grid container item xs={12} sm={6} md={2} lg={1} justify="center">
              <Button
                type="submit"
                variant="contained"
                size="large"
                className="submit-button"
                data-testid="file-submit"
                disabled={sendButtonDisabled}
                onClick={handleFileDataSubmit}
              >
                Submit
              </Button>
            </Grid>
          </Grid>
        </Grid>
      </form>

      <UploadConfirmation isOpen={isConfirmationOpen} closeSnackbar={closeConfirmation} />
    </div>
  );
};

export default UploadVideo;
