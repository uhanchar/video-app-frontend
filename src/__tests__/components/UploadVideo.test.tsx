import React from 'react';
import { screen, fireEvent, queryByAttribute, waitFor } from '@testing-library/react';
import axios from 'axios';
import thunk from 'redux-thunk';
import { applyMiddleware, createStore } from 'redux';

import { render as renderWithRedux } from 'test-utils';
import UploadVideo from 'components/UploadVideo/UploadVideo';
import { RootState } from 'reducers/types';
import rootReducer from 'reducers';
import { CustomAny } from 'constants/data-types';
import { ErrorMessages } from 'constants/error-messages.enum';

jest.mock('axios');
const mockedAxios = axios as jest.Mocked<typeof axios>;

const initialState: RootState = {
  video: {
    video: null,
    isLoading: false,
    error: '',
  },
  app: {
    videos: [],
  },
};

let store: CustomAny;

const getById = queryByAttribute.bind(null, 'id');

const returnEvent = (value: string) => ({
  preventDefault() {},
  target: { value },
});

const returnFileInputEvent = (file: File | null) => ({
  target: {
    files: {
      item: jest.fn().mockImplementationOnce(() => file),
    },
  },
});

describe('test for UploadVideo component', () => {
  beforeEach(() => {
    store = createStore(rootReducer, initialState, applyMiddleware(thunk));
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render Upload Video component', () => {
    const { asFragment } = renderWithRedux(<UploadVideo />, { initialState, store });

    expect(asFragment()).toMatchSnapshot();
  });

  it('should update file title on input', () => {
    const { container } = renderWithRedux(<UploadVideo />, { initialState, store });
    const input = getById(container as HTMLElement, 'video-title');
    const valueInput = 'test-title';

    fireEvent.change(input!, returnEvent(valueInput));

    expect(screen.getByDisplayValue(valueInput)).toBeInTheDocument();
  });

  it('should update file description on input', () => {
    const { container } = renderWithRedux(<UploadVideo />, { initialState, store });
    const input = getById(container as HTMLElement, 'description');
    const valueInput = 'test-description';

    fireEvent.change(input!, returnEvent(valueInput));

    expect(screen.getByDisplayValue(valueInput)).toBeInTheDocument();
  });

  it('should add new file on file input change', () => {
    const { getByTestId } = renderWithRedux(<UploadVideo />, { initialState, store });
    const testFileName = 'test.mpeg4';
    const file = new File([ 'test' ], testFileName, { type: 'video/mpeg4' });
    const fileInput = getByTestId('file-input');

    fireEvent.change(fileInput, returnFileInputEvent(file));

    expect(getByTestId('file-name')).toHaveTextContent(testFileName);
  });

  it('should not change anything if no file exist', () => {
    const { getByTestId } = renderWithRedux(<UploadVideo />, { initialState, store });
    const fileInput = getByTestId('file-input');

    fireEvent.change(fileInput, returnFileInputEvent(null));

    expect(getByTestId('file-name')).toHaveTextContent('');
  });

  it('should show error message if file format is not valid', () => {
    const { getByTestId } = renderWithRedux(<UploadVideo />, { initialState, store });

    const file = new File([ 'test' ], 'test.png', { type: 'image/png' });
    const fileInput = getByTestId('file-input');

    fireEvent.change(fileInput, returnFileInputEvent(file));

    expect(getByTestId('error')).toHaveTextContent(ErrorMessages.WRONG_FILE_FORMAT_MESSAGE);
  });

  it('should send file data on submit button click', async () => {
    mockedAxios.post.mockImplementationOnce(() => Promise.resolve());
    const { getByTestId, container } = renderWithRedux(<UploadVideo />, { initialState, store });

    const file = new File([ 'test' ], 'test.mpeg4', { type: 'video/mpeg4' });
    const fileInput = getByTestId('file-input');
    const fileTitleInput = getById(container as HTMLElement, 'video-title');
    const valueInput = 'test-title';

    fireEvent.change(fileTitleInput!, returnEvent(valueInput));
    fireEvent.change(fileInput, returnFileInputEvent(file));
    fireEvent.click(getByTestId('file-submit'));

    await waitFor(() => getByTestId('close-confirmation'));

    expect(getByTestId('close-confirmation')).toBeInTheDocument();
  });

  it('should show error message if file submit fails', async () => {
    mockedAxios.post.mockImplementationOnce(() => Promise.reject(ErrorMessages.GENERIS_SERVER_ERROR_MESSAGE));
    const { getByTestId, container } = renderWithRedux(<UploadVideo />, { initialState, store });

    const file = new File([ 'test' ], 'test.mpeg4', { type: 'video/mpeg4' });
    const fileInput = getByTestId('file-input');
    const fileTitleInput = getById(container as HTMLElement, 'video-title');
    const valueInput = 'test-title';

    fireEvent.change(fileTitleInput!, returnEvent(valueInput));
    fireEvent.change(fileInput, returnFileInputEvent(file));
    fireEvent.click(getByTestId('file-submit'));

    await waitFor(() => getByTestId('error'));

    expect(getByTestId('error')).toHaveTextContent(ErrorMessages.GENERIS_SERVER_ERROR_MESSAGE);
  });
});
