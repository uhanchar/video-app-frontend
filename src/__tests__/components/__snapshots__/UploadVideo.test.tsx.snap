// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test for UploadVideo component should render Upload Video component 1`] = `
<DocumentFragment>
  <div
    class="upload-video"
  >
    <div
      class="title"
    >
      Upload video
    </div>
    <form
      class="form"
    >
      <div
        class="MuiGrid-root container MuiGrid-container MuiGrid-spacing-xs-3 MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
      >
        <div
          class="MuiGrid-root upload-controls MuiGrid-item MuiGrid-grid-xs-12"
        >
          <div
            class="filename"
            data-testid="file-name"
          />
        </div>
        <div
          class="MuiGrid-root upload-controls MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6 MuiGrid-grid-lg-4"
        >
          <input
            accept="video/*"
            class="input"
            data-testid="file-input"
            id="contained-button-file"
            type="file"
          />
          <label
            for="contained-button-file"
          >
            <span
              aria-disabled="false"
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
              role="button"
              tabindex="0"
            >
              <span
                class="MuiButton-label"
              >
                Upload
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </span>
          </label>
        </div>
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-align-items-xs-center MuiGrid-justify-xs-center MuiGrid-grid-xs-12"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6 MuiGrid-grid-lg-4"
          >
            <div
              class="MuiFormControl-root MuiTextField-root text-input MuiFormControl-fullWidth"
              data-testid="file-title"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
                data-shrink="false"
                for="video-title"
                id="video-title-label"
              >
                Title
                <span
                  aria-hidden="true"
                  class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                >
                   *
                </span>
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input"
                  id="video-title"
                  required=""
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="PrivateNotchedOutline-legendLabelled-3"
                  >
                    <span>
                      Title *
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-align-items-xs-center MuiGrid-justify-xs-center MuiGrid-grid-xs-12"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6 MuiGrid-grid-lg-4"
          >
            <div
              class="MuiFormControl-root MuiTextField-root text-input MuiFormControl-fullWidth"
              data-testid="file-description"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
                data-shrink="false"
                for="description"
                id="description-label"
              >
                Description
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline MuiOutlinedInput-multiline"
              >
                <textarea
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiOutlinedInput-inputMultiline"
                  id="description"
                  rows="1"
                  style="height: 0px; overflow: hidden;"
                />
                <textarea
                  aria-hidden="true"
                  class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiOutlinedInput-inputMultiline"
                  readonly=""
                  style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); width: 100%;"
                  tabindex="-1"
                />
                <fieldset
                  aria-hidden="true"
                  class="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="PrivateNotchedOutline-legendLabelled-3"
                  >
                    <span>
                      Description
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-align-items-xs-center MuiGrid-justify-xs-center MuiGrid-grid-xs-12"
        >
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-justify-xs-center MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-2 MuiGrid-grid-lg-1"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained submit-button MuiButton-containedSizeLarge MuiButton-sizeLarge Mui-disabled Mui-disabled"
              data-testid="file-submit"
              disabled=""
              tabindex="-1"
              type="submit"
            >
              <span
                class="MuiButton-label"
              >
                Submit
              </span>
            </button>
          </div>
        </div>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
